ThreadPool metrics introspection

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
ThreadPool metrics introspection

There are lots of thread pools and simple local-vs-global is not enough already, it is good to know which one in particular uses threads.

Also this will unhide all possible thread pools in ClickHouse, so that user can understand better what can be done in parallel without looking into the sources.

azat committed:
ThreadPool metrics introspection

There are lots of thread pools and simple local-vs-global is not enough
already, it is good to know which one in particular uses threads.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

vdimir commented on src/Processors/Executors/PipelineExecutor.h:
```diff
@@ -67,8 +67,8 @@ class PipelineExecutor
     ConcurrencyControl::SlotPtr single_thread_slot; // slot for single-thread mode to work using executeStep()
-    std::mutex threads_mutex;
-    std::vector<ThreadFromGlobalPool> threads;
+    ThreadPool pool;
```
Doesn't it change behaviour somehow? Was there any reason to use global pool here? @KochetovNicolai

azat replied:
I've decided to do this separately. Rebased without this patch and to fix conflicts.

azat replied:
#48146

vdimir reviewed (approved):
Not sure about changing `PipelineExecutor`. Added metrics LGTM

azat commented:
Lots of failures, but definitely related:

- ASan/MSan - `pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed`
- UBsan - `uninitialized_algorithms.h:538:79: runtime error: reference binding to null pointer of type 'void *'`
- release - `Address: NULL pointer. Access: read. Address not mapped to object` while accessing `system.errors` (thought it is not reproducible), which later cause `underflow` error in DWARF parser.

<details>

```
clickhouse-server: ../nptl/pthread_mutex_lock.c:81: __pthread_mutex_lock: Assertion `mutex->__data.__owner == 0' failed.
```

And in the [core dump from stress tests](https://s3.amazonaws.com/clickhouse-test-reports/47880/0e3b32745b374eb9ab02035a75c1baa4499fa4b6/stress_test__asan_.html):

```
(lldb) p *arg
(pthread_mutex_t) $1 = {
  __data = {
    __lock = 2
    __count = 0
    __owner = 74
    __nusers = 0
    __kind = 0
    __spins = 0
    __elision = 0
    __list = {
      __prev = nullptr
      __next = nullptr
    }
  }
```

MSan report:

```
/var/log/clickhouse-server/clickhouse-server.err.log:2023.03.23 00:42:27.843049 [ 15501 ] {} <Fatal> BaseDaemon: (version 23.3.1.1, build id: 7864BF1A17D6861E4847E35393B891130487D635) (from thread 13154) (query_id: 3b2d1756-d50d-49a9-8187-98c630aa1ccd) (query: SELECT value FROM system.errors WHERE name = 'FUNCTION_THROW_IF_VALUE_IS_NON_ZERO' AND NOT remote) Received signal Segmentation fault (11)
/var/log/clickhouse-server/clickhouse-server.err.log:2023.03.23 00:42:27.859397 [ 15501 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
```

UBsan report:

```
../contrib/llvm-project/libcxx/include/__memory/uninitialized_algorithms.h:538:79: runtime error: reference binding to null pointer of type 'void *'
    #0 0x22f7c850 in void** std::__1::__uninitialized_allocator_copy[abi:v15000]<std::__1::allocator<void*>, void**, void**, void**>(std::__1::allocator<void*>&, void**, void**, void**) build_docker/../contrib/llvm-project/libcxx/include/__memory/uninitialized_algorithms.h:538:5
    #1 0x22f7c695 in std::__1::enable_if<__is_cpp17_forward_iterator<void**>::value, void>::type std::__1::vector<void*, std::__1::allocator<void*>>::__construct_at_end<void**>(void**, void**, unsigned long) build_docker/../contrib/llvm-project/libcxx/include/vector:1031:17
    #2 0x22f7ca59 in DB::ErrorCodes::Error::Error(DB::ErrorCodes::Error const&) build_docker/../src/Common/ErrorCodes.h:36:12
    #3 0x22f734e2 in DB::ErrorCodes::ErrorPair::ErrorPair(DB::ErrorCodes::ErrorPair const&) build_docker/../src/Common/ErrorCodes.h:47:12
    #4 0x22f734e2 in DB::ErrorCodes::ErrorPairHolder::get() build_docker/../src/Common/ErrorCodes.cpp:733:16
    #5 0x2d428c7d in DB::StorageSystemErrors::fillData(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn>>>&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryInfo const&) const build_docker/../src/Storages/System/StorageSystemErrors.cpp:53:52
    #6 0x2d38763f in DB::IStorageSystemOneBlock<DB::StorageSystemErrors>::read(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long) build_docker/../src/Storages/System/IStorageSystemOneBlock.h:54:9
    #7 0x2fbdfe08 in DB::IStorage::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long) build_docker/../src/Storages/IStorage.cpp:135:17
    #8 0x2f062efb in DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&) build_docker/../src/Interpreters/InterpreterSelectQuery.cpp:2448:18
    #9 0x2f05399c in DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::optional<DB::Pipe>) build_docker/../src/Interpreters/InterpreterSelectQuery.cpp:1412:9
    #10 0x2f052495 in DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&) build_docker/../src/Interpreters/InterpreterSelectQuery.cpp:838:5
    #11 0x2f123efa in DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&) build_docker/../src/Interpreters/InterpreterSelectWithUnionQuery.cpp:303:38
    #12 0x2f1253ba in DB::InterpreterSelectWithUnionQuery::execute() build_docker/../src/Interpreters/InterpreterSelectWithUnionQuery.cpp:377:5
    #13 0x2f5ad6b9 in DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/../src/Interpreters/executeQuery.cpp:713:36
    #14 0x2f5a8e6b in DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) build_docker/../src/Interpreters/executeQuery.cpp:1160:30
    #15 0x309929cf in DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:420:24
    #16 0x309b6e95 in DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:2011:9
    #17 0x31e1a17a in Poco::Net::TCPServerConnection::start() build_docker/../base/poco/Net/src/TCPServerConnection.cpp:43:3
    #18 0x31e1b155 in Poco::Net::TCPServerDispatcher::run() build_docker/../base/poco/Net/src/TCPServerDispatcher.cpp:115:20
    #19 0x32087024 in Poco::PooledThread::run() build_docker/../base/poco/Foundation/src/ThreadPool.cpp:188:14
    #20 0x32082814 in Poco::ThreadImpl::runnableEntry(void*) build_docker/../base/poco/Foundation/src/Thread_POSIX.cpp:335:27
    #21 0x7f16b0e12608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
    #22 0x7f16b0d37132 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
```

And in [release](https://s3.amazonaws.com/clickhouse-test-reports/47880/00c4f195916d80007bcf734afed3837b1c0c0d26/stateless_tests__release_.html):

```
2023.03.27 20:04:56.275556 [ 291171 ] {} <Fatal> BaseDaemon: ########################################
2023.03.27 20:04:56.275628 [ 291171 ] {} <Fatal> BaseDaemon: (version 23.3.1.2399, build id: 6BD7454B47BEB16D5DA2155AA5ECA246B895847F) (from thread 5756) (query_id: 35ea113b-5521-41d7-acee-0d099ad08988) (query: select * from system.errors where last_error_message like '%TOPSECRET%';) Received signal Segmentation fault (11)
2023.03.27 20:04:56.275649 [ 291171 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
2023.03.27 20:04:56.275669 [ 291171 ] {} <Fatal> BaseDaemon: Stack trace: 0xe08ce40 0xe08caa7 0x12049306 0x11fb7792 0x13c8e828 0x135b8de8 0x135aa94d 0x135a9ccd 0x1363f2b6 0x136401f4 0x13961cc7 0x1395f26d 0x1472d96c 0x14743119 0x17674334 0x1767555b 0x177fcde7 0x177fa81d 0x7f7825938609 0x7f782585d133
2023.03.27 20:04:56.278803 [ 291171 ] {} <Fatal> BaseDaemon: 2.1. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__memory/construct_at.h:35: void** std::construct_at[abi:v15000]<void*, void*&, void**>(void**, void*&)
2023.03.27 20:04:56.279422 [ 291171 ] {} <Fatal> BaseDaemon: 2.2. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:298: void std::allocator_traits<std::allocator<void*>>::construct[abi:v15000]<void*, void*&, void, void>(std::allocator<void*>&, void**, void*&)
2023.03.27 20:04:56.279447 [ 291171 ] {} <Fatal> BaseDaemon: 2.3. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__memory/uninitialized_algorithms.h:538: void** std::__uninitialized_allocator_copy[abi:v15000]<std::allocator<void*>, void**, void**, void**>(std::allocator<void*>&, void**, void**, void**)
2023.03.27 20:04:56.279471 [ 291171 ] {} <Fatal> BaseDaemon: 2.4. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/vector:1031: std::enable_if<__is_cpp17_forward_iterator<void**>::value, void>::type std::vector<void*, std::allocator<void*>>::__construct_at_end<void**>(void**, void**, unsigned long)
2023.03.27 20:04:56.279489 [ 291171 ] {} <Fatal> BaseDaemon: 2.5. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/vector:1191: vector
2023.03.27 20:04:56.279506 [ 291171 ] {} <Fatal> BaseDaemon: 2. ./build_docker/../src/Common/ErrorCodes.h:36: DB::ErrorCodes::Error::Error(DB::ErrorCodes::Error const&) @ 0xe08ce40 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287631 [ 699 ] {} <Fatal> BaseDaemon: (version 23.3.1.2399, build id: 6BD7454B47BEB16D5DA2155AA5ECA246B895847F) (from thread 291171) Terminate called for uncaught exception:
2023.03.27 20:04:56.287651 [ 699 ] {} <Fatal> BaseDaemon: Code: 465. DB::Exception: underflow: expected bytes 1, got bytes 0. (CANNOT_PARSE_DWARF), Stack trace (when copying this message, always include the lines below):
2023.03.27 20:04:56.287704 [ 699 ] {} <Fatal> BaseDaemon: 
2023.03.27 20:04:56.287722 [ 699 ] {} <Fatal> BaseDaemon: 0. ./build_docker/../src/Common/Exception.cpp:91: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0xe08d7b5 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287746 [ 699 ] {} <Fatal> BaseDaemon: 1. ./build_docker/../contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception<unsigned long, unsigned long>(int, FormatStringHelperImpl<std::type_identity<unsigned long>::type, std::type_identity<unsigned long>::type>, unsigned long&&, unsigned long&&) @ 0xe055f38 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287766 [ 699 ] {} <Fatal> BaseDaemon: 2. ./build_docker/../src/Common/Dwarf.cpp:0: DB::Dwarf::readAttributeSpec(std::basic_string_view<char, std::char_traits<char>>&) @ 0xe07be63 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287785 [ 699 ] {} <Fatal> BaseDaemon: 3. ./build_docker/../src/Common/Dwarf.h:232: DB::Dwarf::findInlinedSubroutineDieForAddress(DB::Dwarf::CompilationUnit const&, DB::Dwarf::Die const&, DB::Dwarf::LineNumberVM const&, unsigned long, std::optional<unsigned long>, std::vector<DB::Dwarf::CallLocation, std::allocator<DB::Dwarf::CallLocation>>&, unsigned long) const::$_8::operator()(DB::Dwarf::Die const&) const::'lambda'(DB::Dwarf::CompilationUnit const&, unsigned long)::operator()(DB::Dwarf::CompilationUnit const&, unsigned long) const @ 0xe087a9d in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287799 [ 699 ] {} <Fatal> BaseDaemon: 4. ./build_docker/../src/Common/Dwarf.cpp:0: bool std::__function::__policy_invoker<bool (DB::Dwarf::Die const&)>::__call_impl<std::__function::__default_alloc_func<DB::Dwarf::findInlinedSubroutineDieForAddress(DB::Dwarf::CompilationUnit const&, DB::Dwarf::Die const&, DB::Dwarf::LineNumberVM const&, unsigned long, std::optional<unsigned long>, std::vector<DB::Dwarf::CallLocation, std::allocator<DB::Dwarf::CallLocation>>&, unsigned long) const::$_8, bool (DB::Dwarf::Die const&)>>(std::__function::__policy_storage const*, DB::Dwarf::Die const&) @ 0xe08750f in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287815 [ 699 ] {} <Fatal> BaseDaemon: 5. ./build_docker/../src/Common/Dwarf.cpp:516: DB::Dwarf::forEachChild(DB::Dwarf::CompilationUnit const&, DB::Dwarf::Die const&, std::function<bool (DB::Dwarf::Die const&)>) const @ 0xe07c224 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287831 [ 699 ] {} <Fatal> BaseDaemon: 6. ./build_docker/../contrib/llvm-project/libcxx/include/__functional/function.h:818: DB::Dwarf::findInlinedSubroutineDieForAddress(DB::Dwarf::CompilationUnit const&, DB::Dwarf::Die const&, DB::Dwarf::LineNumberVM const&, unsigned long, std::optional<unsigned long>, std::vector<DB::Dwarf::CallLocation, std::allocator<DB::Dwarf::CallLocation>>&, unsigned long) const @ 0xe08147d in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287840 [ 699 ] {} <Fatal> BaseDaemon: 7. ./build_docker/../src/Common/Dwarf.cpp:1398: bool std::__function::__policy_invoker<bool (DB::Dwarf::Die const&)>::__call_impl<std::__function::__default_alloc_func<DB::Dwarf::findInlinedSubroutineDieForAddress(DB::Dwarf::CompilationUnit const&, DB::Dwarf::Die const&, DB::Dwarf::LineNumberVM const&, unsigned long, std::optional<unsigned long>, std::vector<DB::Dwarf::CallLocation, std::allocator<DB::Dwarf::CallLocation>>&, unsigned long) const::$_8, bool (DB::Dwarf::Die const&)>>(std::__function::__policy_storage const*, DB::Dwarf::Die const&) @ 0xe08777c in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287863 [ 699 ] {} <Fatal> BaseDaemon: 8. ./build_docker/../src/Common/Dwarf.cpp:516: DB::Dwarf::forEachChild(DB::Dwarf::CompilationUnit const&, DB::Dwarf::Die const&, std::function<bool (DB::Dwarf::Die const&)>) const @ 0xe07c224 in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287875 [ 699 ] {} <Fatal> BaseDaemon: 9. ./build_docker/../contrib/llvm-project/libcxx/include/__functional/function.h:818: DB::Dwarf::findInlinedSubroutineDieForAddress(DB::Dwarf::CompilationUnit const&, DB::Dwarf::Die const&, DB::Dwarf::LineNumberVM const&, unsigned long, std::optional<unsigned long>, std::vector<DB::Dwarf::CallLocation, std::allocator<DB::Dwarf::CallLocation>>&, unsigned long) const @ 0xe08147d in /usr/lib/debug/usr/bin/clickhouse.debug
2023.03.27 20:04:56.287886 [ 699 ] {} <Fatal> BaseDaemon: 10. ./build_docker/../contrib/llvm-project/libcxx/include/vector:537: DB::Dwarf::findLocation(unsigned long, D
```

</details>

azat commented:
>Lots of failures, but definitely related:

I've just missed few ThreadPool's, hence instead of metric, number of threads had been passed.
I've fixed this (and also added one sanity check).

azat commented:
Test failures does not looks related:

>Integration tests (asan) [1/6] — fail: 1, passed: 365, flaky: 0

- `test_keeper_map/test.py::test_create_drop_keeper_map_concurrent` -  #48180

>Integration tests (tsan) [1/6] — fail: 1, passed: 364, flaky: 0

- `test_hedged_requests/test.py::test_stuck_replica` - #48181

tavplubix commented:
It's not clear what happened. Let's check if revert helps

azat commented:
@tavplubix great that you noticed this, this should fix the problem - #48244
